sqlglot/dialects/duckdb.py

```diff
@@ -738,3 +738,34 @@ def withingroup_sql(self, expression: exp.WithinGroup) -> str:
             this = self.sql(expression, "this").rstrip(")")
 
             return f"{this}{expression_sql})"
+
+        def length_sql(self, expression: exp.Length) -> str:
+            arg = expression.this
+
+            # Dialects like BQ and Snowflake also accept binary values as args, so
+            # DDB will attempt to infer the type or resort to case/when resolution
+            if not expression.args.get("binary") or arg.is_string:
+                return self.func("LENGTH", arg)
+
+            if not arg.type:
+                from sqlglot.optimizer.annotate_types import annotate_types
+
+                arg = annotate_types(arg)
+
+            if arg.is_type(*exp.DataType.TEXT_TYPES):
+                return self.func("LENGTH", arg)
+
+            trycast_blob = exp.TryCast(this=arg, to=exp.DataType.build(exp.DataType.Type.VARBINARY))
+            trycast_varchar = exp.TryCast(
+                this=arg, to=exp.DataType.build(exp.DataType.Type.VARCHAR)
+            )
```

Comment:
why do we need trycast vs a simple cast?

Reply:
From my testing both worked so I think either is fine, but since this is a best-effort attempt `exp.TryCast` might handle any conversion errors more graciously.

Reply:
can you test it to see if a try cast is actually needed

Reply:
Right, it will return `NULL`. I've tried a few BQ -> DDB queries with both string and byte expressions, couldn't repro a scenario where the normal cast would break yet. If we also want to error out in such scenarios, then it's another reason to switch to simple casts